### DISCUSSION TITLE
What to try when the calculation fails?

### DISCUSSION THREAD

**Discussion by tommontgomery450** [BoothGroup/Vayesta#163] [Q&A] [answered]
I have a particular use case of a system for which the RHF, CCSD and MP2 seem to converge when I use pyscf on the whole molecule. I have then looked at applying DMET, fragmented (as in the code snippet below). This seems to raise an error with the DIIS matrix being singular. Looking into the main code loop there seem to be two places where this error is checked for. However for this case it seems to happen on the second iteration once the vcorr is updated and the mean field is updated. Here the function `electron_err` raises an error. Below I have provided the complete example. It would be useful to know what to do in such situations e.g. is there a way to see if this is likely due to a bad choice of fragmentation, localisation, lack of setting advanced options in the CCSD solver or any insights that could help us with why this particular example is failing. Thanks, and thanks for a great library!

Below is the code snippet and I've also attached the vlog.txt and the stack trace
```
import pyscf.gto
import pyscf.scf
import vayesta
import vayesta.dmet

xyz_al_f3_co2 = f"""
Al    -3.230182    2.359894    0.097205
F     -3.646509    1.496583   -1.424611
F     -4.734884    3.093588    0.765080
F     -2.575435    1.181353    1.293432
O     -1.995557    3.638514   -0.290018
C     -0.856947    3.555039    0.066321
O      0.283693    3.476012    0.426267
"""

mol = pyscf.gto.Mole()
mol.atom = xyz_al_f3_co2
mol.basis = "sto-3g"
mol.build()

# Hartree-Fock
mf = pyscf.scf.RHF(mol)
mf.kernel()

# DMET
dmet = vayesta.dmet.DMET(mf, solver="CCSD", maxiter=10, diis=True)
with dmet.sao_fragmentation() as f:
    for frag in [[0], 
                 [1], 
                 [2], 
                 [3], 
                 [4, 5, 6]]:
        f.add_atomic_fragment(frag)
    
e = dmet.kernel()
```

[stack_trace.txt](https://github.com/BoothGroup/Vayesta/files/14849075/stack_trace.txt)
[vlog.txt](https://github.com/BoothGroup/Vayesta/files/14849076/vlog.txt)

**Comment by ghb24** [comment, marked as answer]
It appears here as though the chemical potential optimization is failing. I should mention that the DMET module in vayesta is relatively rudimentary, and not actively developed currently. Instead, the EWF module has more support and more robust protocols for the optimization. In addition, the EWF module can do almost everything that the DMET module can do. I would suggest exploring the examples for this: `molecules/15-chemical-potential.py` shows how the chemical potential can be optimized for the democratically partitioned DM as is done in DMET, while `emb.get_dmet_energy(part_cumulant=False)` will return the DMET energy expression in the literature (where `emb` is your embedding object). However, I would argue that the EWF module also contains functionality which will give improved results compared to the traditional DMET approaches.

Please let us know if you have further issues.
